kong/db/migrations/core/005_120_to_130.lua

```diff
@@ -1,3 +1,139 @@
+local cassandra = require "cassandra"
+local cjson = require "cjson.safe"
+
+local function escape_literal(val)
+  if val == ngx.null then
+    return "NULL"
+  end
+
+  local t_val = type(val)
+  if t_val == "number" then
+    return tostring(val)
+  elseif t_val == "string" then
+    return "'" .. tostring((val:gsub("'", "''"))) .. "'"
+  elseif t_val == "boolean" then
+    return val and "TRUE" or "FALSE"
+  end
+  error("don't know how to escape value: " .. tostring(val))
+end
+
+
+local function add_upstream_algorithm_to_postgress(connector)
+  assert(connector:connect_migrations())
+
+  local select_query = [[
+    SELECT * FROM upstreams;
+  ]]
+
+  for row, err in connector:iterate(select_query) do
+    local algorithm
+    if row.hash_on == "none" then
+      algorithm = escape_literal("round-robin")
+    else
+      algorithm = escape_literal("consistent")
+    end
+
+    local id = escape_literal(row.id)
+    local created_at = escape_literal(row.created_at)
+    local name = escape_literal(row.name or ngx.null)
+    local hash_on = escape_literal(row.hash_on or ngx.null)
+    local hash_fallback = escape_literal(row.hash_fallback or ngx.null)
+    local hash_on_header = escape_literal(row.hash_on_header or ngx.null)
+    local hash_fallback_header = escape_literal(row.hash_fallback_header or ngx.null)
+    local hash_on_cookie = escape_literal(row.hash_on_cookie or ngx.null)
+    local hash_on_cookie_path = escape_literal(row.hash_on_cookie_path or ngx.null)
+    local slots = escape_literal(row.slots or ngx.null)
+    local tags = escape_literal(row.tags or ngx.null)
+    local healthchecks_json, err = cjson.encode(row.healthchecks)
+    if err then
+      return err
+    end
+    local healthchecks = escape_literal(healthchecks_json or ngx.null)
+
+    local delete_query = string.format([[
+      DELETE FROM upstreams WHERE id = %s
+    ]], id)
+    local _, err = connector:query(delete_query)
+    if err then
+      return err
+    end
+
+    local insert_query = string.format([[
+      INSERT INTO upstreams(id, created_at, name, hash_on, hash_fallback,
+        hash_on_header, hash_fallback_header, hash_on_cookie,
+        hash_on_cookie_path, slots, healthchecks, tags, algorithm)
+      VALUES(%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
+    ]], id, created_at, name, hash_on, hash_fallback, hash_on_header,
+      hash_fallback_header, hash_on_cookie, hash_on_cookie_path, slots,
+      healthchecks, tags, algorithm)
```

Comment:
Would it be possible to do the Postgres migration using UPDATE, to avoid the window between DELETE and INSERT?

Reply:
(I think it would also make the function a lot shorter, since you won't have to escape and store all fields)